### PR TITLE
Implement wasm target

### PR DESCRIPTION
**Pull request by igoriakovlev**

### CODE REVIEW

File: build.gradle
Comment (sandwwraith) [inline build.gradle:62]: This repository is required for TC aggregate and nightly builds, do not remove it.

File: core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt
Comment (sandwwraith) [inline core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt:71]: I'd rather avoid changing test data if WASM has some problems with float numbers. You may add some `if`s around here if you need to.

File: core/wasmMain/src/kotlinx/serialization/internal/Platform.kt
Comment (sandwwraith) [inline core/wasmMain/src/kotlinx/serialization/internal/Platform.kt:60]: WASM doesn't have non-reified `toTypedArray()`?

File: formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt
Comment (sandwwraith) [inline formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt:17]: Is there an issue with TODO for this test?

File: formats/json/build.gradle
Comment (sandwwraith) [inline formats/json/build.gradle:29]: Why `jsMain` has to be changed?

Comment (sandwwraith) [reply formats/json/build.gradle:29]: Because you moved some files to share them between WASM and JS? Is KMM plugin OK with that? Does import in IDE work correctly?

File: integration-test/build.gradle
Comment (sandwwraith) [inline integration-test/build.gradle:10]: Same here

**Review by sandwwraith — CHANGES_REQUESTED** [review]

File: build.gradle
Comment (igoriakovlev) [reply build.gradle]: Yes, already fixed.

File: formats/json/build.gradle
Comment (igoriakovlev) [reply formats/json/build.gradle:29]: I do not see any issues to do so. But this commit is not necessary and we can drop it (but we need to copy-paste this files into different targets). Or we can try make new source set and depend on it from target source sets, but then we need to ignore metadata tasks for it.

File: integration-test/build.gradle
Comment (igoriakovlev) [reply integration-test/build.gradle]: Yeah, i already rolled it back. Thanks

File: core/wasmMain/src/kotlinx/serialization/internal/Platform.kt
Comment (igoriakovlev) [reply core/wasmMain/src/kotlinx/serialization/internal/Platform.kt]: It has. Fixed.

File: core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt
Comment (igoriakovlev) [reply core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt:71]: It is quite random numbers. The new ones is not better or worse than old ones :)
You think that it is nice to have here?

File: formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt
Comment (sandwwraith) [reply formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt:17]: Please take a look

File: core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt
Comment (sandwwraith) [reply core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt:71]: If they're random, why change them?

File: formats/json/build.gradle
Comment (igoriakovlev) [reply formats/json/build.gradle:29]: Its a bug in IDE. The only thing we can do is copy-paste the code. And this is only inspection bug (the gutter works fine).

File: core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt
Comment (igoriakovlev) [reply core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt:71]: Agggh...because this data is not suitable to wasm due to lack of wasm stdlib implementation. If i will add another dataset only for wasm how can it make things better?

Comment (pdvrieze) [reply core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt:71]: If the "random" values break something we got lucky that they detect a bug. They shouldn't be changed. Rather the bug fixed, or the test "ignored" if that is best.

File: formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt
Comment (igoriakovlev) [reply formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt]: This seems to be a bug in okio. I keep this test ingored for wasm and put the [link](https://github.com/square/okio/issues/1333) to the issue nearby.

File: formats/json/build.gradle
Comment (sandwwraith) [reply formats/json/build.gradle:32]: I wanna see this IDE bug fixed :) So please report it, if it is not created yet.
IDE consistency and correctness directly affects daily work, so yes, I do not want to see any red code in the IDE if it can be compiled correctly. Alternatively, it is possible to use `@Suppress("NO_ACTUAL_FOR_EXPECT")` but I'm not sure that suppressing errors is better than copy pasting several files. @qwwdfsad Maybe you have an opinion on this?

File: formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt
Comment (sandwwraith) [reply formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt]: Great, thank you!

File: core/commonTest/src/kotlinx/serialization/BasicTypesSerializationTest.kt
Comment (sandwwraith) [inline core/commonTest/src/kotlinx/serialization/BasicTypesSerializationTest.kt:178]: Please add a comment with YouTrack issue for this wasm bug with float numbers

File: core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt
Comment (sandwwraith) [reply core/commonTest/src/kotlinx/serialization/UmbrellaTypes.kt:92]: So old ones ARE better, given that they catch bugs, aren't they? :) 

Good idea with two tests!

File: formats/json/build.gradle
Comment (pdvrieze) [reply formats/json/build.gradle:32]: I have found that in the latest version actuals in a shared module do work, but they require correct dependencies (the shared module must depend on the common module). A case that still breaks is where you want to share code between jvm and android and use (shared) jvm types.

Comment (igoriakovlev) [reply formats/json/build.gradle:32]: I see that this invalid report occurs on current `dev` branch (without wasm target added), so it seems it is not a result of my changes.

Comment (igoriakovlev) [reply formats/json/build.gradle:32]: https://youtrack.jetbrains.com/issue/KTIJ-26668/False-positive-diagnostic-for-missing-actual-declararion-for-expect-object

Comment (igoriakovlev) [reply formats/json/build.gradle:32]: So what do you prefer?

- Copy and Paste (does not work anyway, the diagnostic will appear)
- Directory sharing
- Sourceset dependency

Comment (sandwwraith) [reply formats/json/build.gradle:32]: I'm on `dev` and I don't see this error. Did you reimport project after branch switching?

Comment (igoriakovlev) [reply formats/json/build.gradle:32]: I made new project.

File: core/commonTest/src/kotlinx/serialization/BasicTypesSerializationTest.kt
Comment (sandwwraith) [reply core/commonTest/src/kotlinx/serialization/BasicTypesSerializationTest.kt]: Why did you delete added tests though?

File: formats/json/build.gradle
Comment (sandwwraith) [reply formats/json/build.gradle:32]: How it is related to serializtion then? With only jsMain, this error doesn't happen.

We'll discuss it on the next week, since now it is time to release 1.6.0 (this PR won't go there anyway)

Comment (igoriakovlev) [reply formats/json/build.gradle:32]: Well, as I see, with my IDEA it does not related at all.
And I see that with any other MPP project (even without wasm).
I dont understand why it so bother you, its a bug and it is reported.

Screenshot 2023-08-17 at 15 55 17

File: core/commonTest/src/kotlinx/serialization/BasicTypesSerializationTest.kt
Comment (igoriakovlev) [reply core/commonTest/src/kotlinx/serialization/BasicTypesSerializationTest.kt]: I delete only one that was originally ok. There is two similar test with similar testdata. One is passing and the second is not.

File: formats/json/build.gradle
Comment (igoriakovlev) [reply formats/json/build.gradle:32]: I am trying understand the reasons for all of that....if tomorrow new bugs will appear in IDEA, would you rewrite your code to escape this bug or what? Or stop to getting MR's that affected with this bug in editor? What is going on? Why I am wasting my (and yours time). If you are so aware for editor problems with green code, push the IDEA but not codeowners.

File: formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt
Comment (swankjesse) [reply formats/json-tests/commonTest/src/kotlinx/serialization/features/EmojiTest.kt]: https://youtrack.jetbrains.com/issue/KT-60212

File: formats/json/build.gradle
Comment (sandwwraith) [reply formats/json/build.gradle:32]: It's not a stopper for the PR, indeed. I just want to find a suitable workaround.

Comment (igoriakovlev) [reply formats/json/build.gradle:32]: Well, I discussed this with a guys from MPP and we seems to be found whats wrong. Its `.idea` directory that contains artefacts that infer with compiler settings. `.idea` directory does not deleted by any recoveries or cache cleaning options of IDEA, that is why diagnostics persists. You can remove `.idea` folder and this error should away.
We can also compare old `.idea` files and new ones and found the cause (i have already deleted this directory so cannot compare it.

Comment (sandwwraith) [reply formats/json/build.gradle:32]: Great find, thanks! I'll take a look at .idea

Comment (sandwwraith) [reply formats/json/build.gradle:32]: I don't know what changed exactly, but the error seems gone for me after switching branches. I guess we can mark it as resolved for now, applying `@Suppress` if it happens again. As I said, I'll merge this PR after 1.6.0 release (tomorrow)

Comment (igoriakovlev) [reply formats/json/build.gradle:32]: Perfect. Thanks. Is it ok for you current approach with intermediate `sourceSet` and ignoring `compileJsWasmMainKotlinMetadata` task as I do for now?

Comment (sandwwraith) [reply formats/json/build.gradle:32]: > ignoring compileJsWasmMainKotlinMetadata task

Can you elaborate on that? I don't see any changes in buildfiles relating to this task

File: formats/json/build.gradle
Comment (sandwwraith) [inline formats/json/build.gradle:44]: Ah, found it. Can you add a comment here explaining why it is needed?

Comment (igoriakovlev) [reply formats/json/build.gradle:56]: Done

**Comment by igoriakovlev** [comment]
I have added commit to make possible build and test with kotlin 1.9.30+

**Review by swankjesse — COMMENTED** [review]
LGTM

**Review by sandwwraith — APPROVED** [review]

**Comment by sandwwraith** [comment]
Great job! 💪🏻

**Comment by igoriakovlev** [comment]
Thank you all for the review!

**Pull request merged**
